From c3bc2b1f4205b7780cc29d9ac4b7636cdcd87847 Mon Sep 17 00:00:00 2001
From: Rajmohan Mani <rajmohan.mani@intel.com>
Date: Thu, 5 Apr 2018 17:39:18 -0700
Subject: [PATCH 017/129] media: intel-ipu4: isys: Compile meta data code
 selectively

Message for Open Source:
Added compiler flag to selectively compile the meta data code.

Message for Internal:
[Issue/Feature]
Certain platforms do not need meta data code.
Currently there is no way to compile this code
selectively for platforms where it is needed.
[Root Cause/Changes]
Added IPU_META_DATA_SUPPORT compiler flag to selectively
compile the meta data code.
Added this compiler flag in respective Makefiles,
to enable this code by default.

Change-Id: I0dbf7593deb21309427f4f6cd0eebccd57156126
Depends-On: I9d054c390b5c7bbd0ca14a611b4d79faf518bedf
Tracked-On: #JICCCN-130
Signed-off-by: Rajmohan Mani <rajmohan.mani@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/pci/intel/ipu-isys-csi2.c | 37 ++++++++++++++++++++++++++-------
 drivers/media/pci/intel/ipu-isys-csi2.h |  6 ++++++
 drivers/media/pci/intel/ipu4/Makefile   |  2 ++
 3 files changed, 37 insertions(+), 8 deletions(-)

diff --git a/drivers/media/pci/intel/ipu-isys-csi2.c b/drivers/media/pci/intel/ipu-isys-csi2.c
index ff1fe0a..93e27f0 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2.c
+++ b/drivers/media/pci/intel/ipu-isys-csi2.c
@@ -75,10 +75,12 @@
 	0,
 };
 
+#ifdef IPU_META_DATA_SUPPORT
 static const u32 csi2_supported_codes_pad_meta[] = {
 	MEDIA_BUS_FMT_FIXED,
 	0,
 };
+#endif
 
 static const u32 *csi2_supported_codes[NR_OF_CSI2_PADS];
 
@@ -128,6 +130,7 @@ int ipu_isys_csi2_get_link_freq(struct ipu_isys_csi2 *csi2, __s64 *link_freq)
 	return 0;
 }
 
+#ifdef IPU_META_DATA_SUPPORT
 static int ipu_get_frame_desc_entry_by_dt(struct v4l2_subdev *sd,
 					  struct v4l2_mbus_frame_desc_entry
 					  *entry, u8 data_type)
@@ -187,6 +190,7 @@ static void csi2_meta_prepare_firmware_stream_cfg_default(
 		    entry.size.two_dim.height;
 	}
 }
+#endif
 
 static int subscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
 			   struct v4l2_event_subscription *sub)
@@ -206,11 +210,12 @@ static int subscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
 	.unsubscribe_event = v4l2_event_subdev_unsubscribe,
 };
 
+#ifdef IPU_META_DATA_SUPPORT
 struct ipu_isys_pixelformat csi2_meta_pfmts[] = {
 	{V4L2_FMT_IPU_ISYS_META, 8, 8, 0, MEDIA_BUS_FMT_FIXED, 0},
 	{},
 };
-
+#endif
 /*
  * The input system CSI2+ receiver has several
  * parameters affecting the receiver timings. These depend
@@ -431,10 +436,10 @@ static int csi2_link_validate(struct media_link *link)
 bool csi2_has_route(struct media_entity *entity, unsigned int pad0,
 		    unsigned int pad1, int *stream)
 {
-	/* TODO: need to remove this when meta data node is removed */
+#ifdef IPU_META_DATA_SUPPORT
 	if (pad0 == CSI2_PAD_META || pad1 == CSI2_PAD_META)
 		return true;
-
+#endif
 	return ipu_isys_subdev_has_route(entity, pad0, pad1, stream);
 }
 
@@ -442,6 +447,7 @@ bool csi2_has_route(struct media_entity *entity, unsigned int pad0,
 	.s_stream = set_stream,
 };
 
+#ifdef IPU_META_DATA_SUPPORT
 static int get_metadata_fmt(struct v4l2_subdev *sd,
 			    struct v4l2_subdev_pad_config *cfg,
 			    struct v4l2_subdev_format *fmt)
@@ -468,14 +474,16 @@ static int get_metadata_fmt(struct v4l2_subdev *sd,
 	}
 	return rval;
 }
+#endif
 
 static int ipu_isys_csi2_get_fmt(struct v4l2_subdev *sd,
 				 struct v4l2_subdev_pad_config *cfg,
 				 struct v4l2_subdev_format *fmt)
 {
+#ifdef IPU_META_DATA_SUPPORT
 	if (fmt->pad == CSI2_PAD_META)
 		return get_metadata_fmt(sd, cfg, fmt);
-
+#endif
 	return ipu_isys_subdev_get_ffmt(sd, cfg, fmt);
 }
 
@@ -483,9 +491,10 @@ static int ipu_isys_csi2_set_fmt(struct v4l2_subdev *sd,
 				 struct v4l2_subdev_pad_config *cfg,
 				 struct v4l2_subdev_format *fmt)
 {
+#ifdef IPU_META_DATA_SUPPORT
 	if (fmt->pad == CSI2_PAD_META)
 		return get_metadata_fmt(sd, cfg, fmt);
-
+#endif
 	return ipu_isys_subdev_set_ffmt(sd, cfg, fmt);
 }
 
@@ -547,6 +556,7 @@ static void csi2_set_ffmt(struct v4l2_subdev *sd,
 		return;
 	}
 
+#ifdef IPU_META_DATA_SUPPORT
 	if (fmt->pad == CSI2_PAD_META) {
 		struct v4l2_mbus_framefmt *ffmt =
 			__ipu_isys_get_ffmt(sd, cfg, fmt->pad,
@@ -579,7 +589,7 @@ static void csi2_set_ffmt(struct v4l2_subdev *sd,
 
 		return;
 	}
-
+#endif
 	if (sd->entity.pads[fmt->pad].flags & MEDIA_PAD_FL_SOURCE) {
 		ffmt->width = fmt->format.width;
 		ffmt->height = fmt->format.height;
@@ -628,7 +638,10 @@ void ipu_isys_csi2_cleanup(struct ipu_isys_csi2 *csi2)
 	ipu_isys_subdev_cleanup(&csi2->asd);
 	for (i = 0; i < NR_OF_CSI2_SOURCE_PADS; i++)
 		ipu_isys_video_cleanup(&csi2->av[i]);
+	
+#ifdef IPU_META_DATA_SUPPORT
 	ipu_isys_video_cleanup(&csi2->av_meta);
+#endif
 	csi2->isys = NULL;
 }
 
@@ -662,10 +675,12 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 			   .height = 3072,
 			  },
 	};
+#ifdef IPU_META_DATA_SUPPORT
 	struct v4l2_subdev_format fmt_meta = {
 		.which = V4L2_SUBDEV_FORMAT_ACTIVE,
 		.pad = CSI2_PAD_META,
 	};
+#endif
 	int i, rval, src;
 
 	csi2->isys = isys;
@@ -694,7 +709,9 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 	     i < (NR_OF_CSI2_SOURCE_PADS + CSI2_PAD_SOURCE(0)); i++)
 		csi2->asd.pad[i].flags = MEDIA_PAD_FL_SOURCE;
 
+#ifdef IPU_META_DATA_SUPPORT
 	csi2->asd.pad[CSI2_PAD_META].flags = MEDIA_PAD_FL_SOURCE;
+#endif
 	src = index;
 #ifdef CONFIG_VIDEO_INTEL_IPU4P
 	src = index ? (index + 5) : (index + 3);
@@ -704,7 +721,9 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 
 	for (i = 0; i < NR_OF_CSI2_SOURCE_PADS; i++)
 		csi2_supported_codes[i + 1] = csi2_supported_codes_pad_source;
+#ifdef IPU_META_DATA_SUPPORT
 	csi2_supported_codes[CSI2_PAD_META] = csi2_supported_codes_pad_meta;
+#endif
 	csi2->asd.supported_codes = csi2_supported_codes;
 	csi2->asd.set_ffmt = csi2_set_ffmt;
 
@@ -723,7 +742,9 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 	}
 
 	__ipu_isys_subdev_set_ffmt(&csi2->asd.sd, NULL, &fmt);
+#ifdef IPU_META_DATA_SUPPORT
 	__ipu_isys_subdev_set_ffmt(&csi2->asd.sd, NULL, &fmt_meta);
+#endif
 
 	/* create default route information */
 	for (i = 0; i < NR_OF_CSI2_STREAMS; i++) {
@@ -776,7 +797,7 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 		}
 	}
 
-	/* TODO: remove meta data pad */
+#ifdef IPU_META_DATA_SUPPORT
 	snprintf(csi2->av_meta.vdev.name, sizeof(csi2->av_meta.vdev.name),
 		 IPU_ISYS_ENTITY_PREFIX " CSI-2 %u meta", index);
 	csi2->av_meta.isys = isys;
@@ -803,7 +824,7 @@ int ipu_isys_csi2_init(struct ipu_isys_csi2 *csi2,
 		dev_info(&isys->adev->dev, "can't init metadata node\n");
 		goto fail;
 	}
-
+#endif
 	return 0;
 
 fail:
diff --git a/drivers/media/pci/intel/ipu-isys-csi2.h b/drivers/media/pci/intel/ipu-isys-csi2.h
index c4b7ce0..34c202f 100644
--- a/drivers/media/pci/intel/ipu-isys-csi2.h
+++ b/drivers/media/pci/intel/ipu-isys-csi2.h
@@ -25,10 +25,14 @@
 	(__n >= NR_OF_CSI2_SOURCE_PADS ? \
 		(NR_OF_CSI2_PADS - 2) : \
 		(__n + NR_OF_CSI2_SINK_PADS)); })
+#ifdef IPU_META_DATA_SUPPORT
 #define NR_OF_CSI2_META_PADS		1
 #define NR_OF_CSI2_PADS			\
 	(NR_OF_CSI2_SINK_PADS + NR_OF_CSI2_SOURCE_PADS + NR_OF_CSI2_META_PADS)
 #define CSI2_PAD_META			(NR_OF_CSI2_PADS - 1)
+#else
+#define NR_OF_CSI2_PADS	(NR_OF_CSI2_SINK_PADS + NR_OF_CSI2_SOURCE_PADS)
+#endif
 
 #define IPU_ISYS_SHORT_PACKET_BUFFER_NUM	VIDEO_MAX_FRAME
 #define IPU_ISYS_SHORT_PACKET_WIDTH	32
@@ -90,7 +94,9 @@ struct ipu_isys_csi2 {
 	struct ipu_isys *isys;
 	struct ipu_isys_subdev asd;
 	struct ipu_isys_video av[NR_OF_CSI2_SOURCE_PADS];
+#ifdef IPU_META_DATA_SUPPORT
 	struct ipu_isys_video av_meta;
+#endif
 	struct completion eof_completion;
 
 	void __iomem *base;
diff --git a/drivers/media/pci/intel/ipu4/Makefile b/drivers/media/pci/intel/ipu4/Makefile
index 7a20874..1278146 100644
--- a/drivers/media/pci/intel/ipu4/Makefile
+++ b/drivers/media/pci/intel/ipu4/Makefile
@@ -7,6 +7,7 @@ endif
 
 ifdef CONFIG_VIDEO_INTEL_IPU4
 ccflags-y += -DHAS_DUAL_CMD_CTX_SUPPORT=0
+ccflags-y += -DIPU_META_DATA_SUPPORT
 
 intel-ipu4-objs				+= ../ipu.o \
 					   ../ipu-bus.o \
@@ -69,6 +70,7 @@ endif
 
 ifdef CONFIG_VIDEO_INTEL_IPU4P
 ccflags-y += -DHAS_DUAL_CMD_CTX_SUPPORT=0
+ccflags-y += -DIPU_META_DATA_SUPPORT
 
 intel-ipu4p-objs			+= ../ipu.o \
 					   ../ipu-bus.o \
-- 
1.9.1

